src/codegen/encoding/borsh_encoding.rs

```diff
@@ -0,0 +1,624 @@
+use crate::ast::{Namespace, RetrieveType, Type};
+use crate::codegen::cfg::{ControlFlowGraph, Instr};
+use crate::codegen::encoding::{
+    calculate_size_args, finish_array_loop, increment_four, load_array_item, load_struct_member,
+    load_sub_array, set_array_loop, Encoding,
+};
+use crate::codegen::vartable::Vartable;
+use crate::codegen::{Builtin, Expression};
+use num_bigint::BigInt;
+use num_traits::{One, Zero};
+use solang_parser::pt::Loc;
+use std::collections::HashMap;
+use std::ops::MulAssign;
+
+/// This struct implements the trait Encoding for Borsh encoding
+pub(super) struct BorshEncoding {
+    /// The trait Encoding has a 'cache_storage_loaded' function, which saves the variables to which
+    /// we loaded items from storage. This attribute is map between the argument number (in the
+    /// 'abi.encode' function) to the expression that represents the item loaded from storage.
```

Comment:
please explain why we're keeping a cache of these

Reply:
"caching avoids having to repeat the same operation" is just not a useful comment.

Explain *why* you are caching things. Think of the 5 whys. Asking another why.

So, explain that why you need to do multiple passes over the data. 
- One to calculate the length, then an allocate with the size, and then the actual encoding.
-  Explain what's in the expression in the cache. Is it a clone of the original expression or are you storing the result in a `Expression::Variable(..)`.
- Explain what operations you are caching and which operations you are not caching
- The name suggests this is only a cache for arguments which are in storage, not memory.
- Why are you not caching other results. `abi.encode()` can sometimes have expensive arguments like `keccak256()`: https://github.com/hyperledger-labs/solang/blob/main/integration/substrate/UniswapV2ERC20.sol#L27-L33

Reply:
> The name suggests this is only a cache for arguments which are in storage, not memory.

The name is correct. I'm only caching arguments that have been loaded from storage.

> Why are you not caching other results. abi.encode() can sometimes have expensive arguments like keccak256(): https://github.com/hyperledger-labs/solang/blob/main/integration/substrate/UniswapV2ERC20.sol#L27-L33

There is no need to cache `keccka256`, because it is only calculated once. We know its size during compile time (256 bits), so the `get_expr_size` will return 32 bytes directly, without evaluating the result.

Reply:
I updated the comment both in `encoding/mod.rs` and `borsh_encoding.rs`.